CRM/Custom/Import/Form/MapField.php

```diff
@@ -72,16 +74,13 @@ public static function formRule($fields) {
       foreach ($fields['mapper'] as $mapperPart) {
         $importKeys[] = $mapperPart[0];
       }
-      $requiredFields = array(
-        'contact_id' => ts('Contact ID'),
-      );
-      foreach ($requiredFields as $field => $title) {
-        if (!in_array($field, $importKeys)) {
-          if (!isset($errors['_qf_default'])) {
-            $errors['_qf_default'] = '';
-          }
-          $errors['_qf_default'] .= ts('Missing required field: %1', array(1 => $title));
+
+      // check either contact id or external identifier
+      if (!in_array('contact_id', $importKeys) && !in_array('external_identifier', $importKeys)) {
+        if (!isset($errors['_qf_default'])) {
+          $errors['_qf_default'] = '';
         }
+        $errors['_qf_default'] .= ts('Missing required field: %1', array(1 => ts('Contact ID')));
```

Comment:
I changed "upstream" by "customfields", but for the rest I followed your steps and it should be ok now :-)